lock-api/src/main/java/com/palantir/lock/client/MultiClientBatchingIdentifiedAtlasDbTransactionStarter.java

```diff
@@ -0,0 +1,319 @@
+                        .safeLoggablePurpose("multi-client-transaction-starter")
+                        .build();
+        return new MultiClientBatchingIdentifiedAtlasDbTransactionStarter(autobatcher);
+    }
+
+    public List<StartIdentifiedAtlasDbTransactionResponse> startTransactions(
+            Namespace namespace,
+            int request,
+            StartTransactionsLockWatchEventCache cache,
+            LockLeaseService.LockCleanupService lockCleanupService) {
+        return AtlasFutures.getUnchecked(autobatcher.apply(
+                NamespaceAndRequestParams.of(namespace, RequestParams.of(request, cache, lockCleanupService))));
+    }
+
+    private static Consumer<
+                    List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>>>
+            consumer(InternalMultiClientConjureTimelockService delegate, UUID requestorId) {
+        return batch -> processBatch(delegate, requestorId, batch);
+    }
+
+    @VisibleForTesting
+    static void processBatch(
+            InternalMultiClientConjureTimelockService delegate,
+            UUID requestorId,
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+
+        Map<Namespace, ResponseHandler> namespaceWiseResponseHandler = getResponseHandlers(batch);
+        MultiClientRequestManager multiClientRequestManager =
+                new MultiClientRequestManager(getNamespaceWiseRequestParams(batch));
+
+        Map<Namespace, List<StartIdentifiedAtlasDbTransactionResponse>> startTransactionResponses = new HashMap<>();
+
+        while (multiClientRequestManager.requestsPending()) {
+            try {
+                startTransactionResponses =
+                        getStartTransactionResponses(multiClientRequestManager.requestMap, delegate, requestorId);
+                startTransactionResponses.forEach((namespace, responseList) -> {
+                    multiClientRequestManager.updatePendingStartTransactionsCount(namespace, responseList.size());
+                    namespaceWiseResponseHandler.get(namespace).processResponse(responseList);
+                });
+            } catch (Throwable t) {
+                clearResources(multiClientRequestManager, startTransactionResponses);
+                throw Throwables.throwUncheckedException(t);
+            }
+        }
+    }
+
+    private static void clearResources(
+            MultiClientRequestManager multiClientRequestManager,
+            Map<Namespace, List<StartIdentifiedAtlasDbTransactionResponse>> startTransactionResponses) {
+        startTransactionResponses.forEach((namespace, responseList) -> TransactionStarterHelper.unlock(
+                responseList.stream()
+                        .map(response -> response.immutableTimestamp().getLock())
+                        .collect(Collectors.toSet()),
+                multiClientRequestManager.getLockCleanupService(namespace)));
+    }
```

Comment:
I think we also need to clear pending responses in the multi-client request manager?

Reply:
It may be easier to have it be (auto)closeable and have a finally block around the while loop in `processBatch`.